all/*/board.json: Simplify board feature tags.

jimmo opened the pull request:
- Find a common set of board feature tags and map existing features to that reduced set.
- Remove some less-useful board feature tags.
- Ensure all mcus are specified correctly.
- Ensure all boards have a vendor (and fix some vendor names).

This is to make the downloads page show a less intimidating set of filters.

Work done in conjunction with @mattytrentini .

_This work was funded through GitHub Sponsors._

jimmo committed:
ports: Simplify board feature tags in board.json.

This commit:
- Finds a common set of board feature tags and maps existing features to
  that reduced set.
- Removes some less-useful board feature tags.
- Ensures all MCUs are specified correctly.
- Ensures all boards have a vendor (and fixes some vendor names).

This is to make the downloads page show a less intimidating set of filters.

Work done in conjunction with Matt Trentini <matt.trentini@gmail.com>.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
The bulk of this PR was done by a script that applied the following:

```py
RENAME_FEATURES = {
    "10/100 Ethernet Phy": ["Ethernet"],
    "8MB SDRAM": ["External RAM"],
    "AudioAmp": ["Audio"],
    "AudioCodec": ["Audio"],
    "Battery Management": ["Battery Charging"],
    "CYW43 WiFi/BT Module": ["WiFi", "BLE"],
    "Grove": ["JST-PH"],
    "Humidity sensor HTS221": ["Environment Sensor"],
    "Micro USB": ["USB"],
    "MicroSD": ["microSD"],
    "MicroUSB": ["USB"],
    "OLED": ["Display"],
    "Pressure sensor LPS22H": ["Environment Sensor"],
    "Proximity, Light, RGB sensor APDS-9960": ["Environment Sensor"],
    "QWIIC": ["JST-SH"],
    "Red/green/orange/blue leds": ["RGB LED"],
    "SDRAM": ["External RAM"],
    "SPI Flash 16MB": ["External RAM"],
    "SPIRAM": ["External RAM"],
    "STEMMA QT/QWIIC": ["JST-SH"],
    "USB-A": ["USB"],
    "USB-MICRO": ["USB"],
    "WiFi Nina-W102": ["WiFi"],
    "USB Full speed": ["USB"],
    "USB High Speed Phy": ["USB"],
    "USB Stick form factor": ["USB"],
}

REMOVE_FEATURES = {
    "16MB Flash",
    "1MB SPI Flash",
    "2x80 pin HD connectors",
    "ADC",
    "Breadboard Friendly",
    "Breadboard friendly",
    "Castellated Pads",
    "DisplayPort over USB-C",
    "Dual-core processor",
    "I2C",
    "Infrared",
    "JLink",
    "OpenSDA",
    "PWM",
    "QSPI Flash",
    "Reset/User button",
    "SIM Socket",
    "SPDIF",
    "SPI Flash",
    "SPI",
    "UART",
}
```

The only other changes were manual modifications to vendor and MCU.

jimmo commented:
@mattytrentini As discussed I think we should add a CI action to "lint" the board.json files (but in a separate PR).

The idea being that adding a new feature tag is fine, but should also include a corresponding change to the linter to add it to the "allowed" set. Also verify that every new board.json meets basic criteria (e.g. has a vendor, url, image, etc).

jimmo commented:
There are a couple of removed feature tags that, rather than removing them, we could consider moving to the optional "features_non_filterable" field in board.json (essentially this is a free-form that the board owner can use to add additional list content to the board's downloads page).

robert-hh commented:
This rename seems wrong: ` "SPI Flash 16MB": ["External RAM"],`
Besides that you removed the tag for external Flash. But that may be an important feature, used e.g. for a larger file system.

jimmo commented:
> This rename seems wrong: ` "SPI Flash 16MB": ["External RAM"],` Besides that you removed the tag for external Flash. But that may be an important feature, used e.g. for a larger file system.

@robert-hh Thanks, will fix.

We were on the fence about external flash. I agree with you about it being relevant for a larger filesystem, but I'm also a little bit unsure about micros that don't have internal flash. i.e. does an ESP8266 with 512kiB flash count as "External Flash"? Or is it only "external flash" if it's not also used for firmware? Or above a certain size?

Maybe the nuance here doesn't matter -- anything that has SPI flash in any form gets the "External flash" tag.

I'm happy either way.

jimmo commented:
Thinking about it some more, I think for now we should keep "External Flash" and add it on all rp2040,esp8266,esp32,mimxrt boards that don't already have it.
We can remove it later in different PR if necessary.

jimmo commented:
Restored "External Flash" (replacing all the variants of "SPI Flash") and added for all rp2040,esp8266,esp32,mimxrt boards
 
Also added "Dual-core" for rp2040 and the three stm32h747 boards. This is a bit tricky, for example you could argue that it should also be added for the ESP32, but dual-core is not supported by MicroPython on ESP32. Although technically neither is it supported on the H7, and less likely to be supported anytime soon compared to ESP32. My argument is that it's a unique feature of the H747, and a well-supported feature of the rp2040.

robert-hh commented:
Thanks. This is quite a challenge. Looking at the MIMXRT and SAMD series, I've seen a few aspects, which surely were wrong before.
SAMD: 
- The Seeed WIO terminal has External Flash.
MIMXRT: 
- I do not know what you call audio. But all MIMXRTxxxx_EVK boards have a codec on-board, a Microphone, a headphone jack and SPDIF support prepared (RSA plug & transformer missing)
- The MIMXRT1015 board has no MicroSD-Card slot and no CAN support hardware (my fault).
- The MIMXRT1050 has like the MIMXRT1060/64 an 76 and the Seeed Arch Mix a connector for camera and display, but it's not supported by MicroPython.
- The Olimex RT1010 has not Audio Codec (my fault).
- The Teensy 4.0 does not have a MicroSD slot, but obviously it has USB.
- The Teensy 4.1 has USB.


I do not know whether that should be changed now, or if you prefer to have the online first and then I could make a PR with updates.

jimmo commented:
Thanks @robert-hh 

I think this is worth fixing now -- so I've made the fixes you suggested.

"Audio" was a catch-all term for a bunch of things, but I think "Audio Codec" makes more sense, so I've updated the mimxrt boards to use that. Some boards have microphones without audio codecs, so I've left "Microphone" as a separate tag.

It's a bit fuzzy to me what types of connectors we should add tags for. e.g. JST seems like a obvious one. I don't think we need to worry about the camera/display connector on the mimxrt boards for now. Not sure about headphone jack, but leaning towards leaving it out for now.

dpgeorge commented:
Thanks for this, a much needed clean up.
